labs/b5.md

```diff
@@ -0,0 +1,364 @@
+This layer deals with transfer data between network nodes in a wide area network (WAN) or a local area network (LAN). An example of this is L2 routing according to MAC addresses.
+3. **Network Layer**
+This layer deals with packet forwarding and routing through intermediate routers. The most common L3 protocol, and the one you're probably familiar with, is Internet Protocol (IP).  This layer is concerned with delivering data between hosts that correspond to IP addresses but makes no guarantees about reliable transport of the packets.
+4. **Transport Layer**
+This layer is responsible for providing reliable transport, and reassembling packets that may have arrived out of order.  Protocols in this layer provide host-to-host communication services for applications. The most well known protocols are TCP (connection-oriented) and UDP(connection-less).
+5. **Session Layer**
+6. **Presentation Layer**
+Won't talk much about the above two, they aren't as important and are sometimes not even implemented in a network stack.
+7. **Application Layer**
+The application layer specifies protocols between hosts. Some examples of this include SSH, FTP, and HTTP. 
+
+Here is a diagram of the OSI model
+![alt text][OSI]
+
+[OSI]: https://i-technet.sec.s-msft.com/dynimg/IC213395.gif "OSI model"
+
+## L2 Routing
+Link layer routing, aka L2 routing, is based off MAC addresses and uses a switch to forward frames to the proper MAC.
+
+## Switches
+Switches are physical devices that connect devices on a network and use packet switching to receive, process, and forward data.  Switches process this data at L2 and therefore rely on MAC addresses to identify hosts and place them in the network.
+
+## MAC
+Media access control (MAC) addresses are a identifiers uniquely assigned to network interfaces.
+![alt text][MAC]
+
+[MAC]: https://upload.wikimedia.org/wikipedia/commons/9/94/MAC-48_Address.svg "MAC address format"
+
+Since the MAC address is unique this is often referred to as the physical address. The octets are often written in hexadecimal and delimited by colons. An example MAC address is  `00:14:22:01:23:45`. Note that the first 3 octets refer to the Organizationally Unique Identifier (OUI) which can help identify manufacturers.  Fun fact -- the `00:14:22` above is an OUI for Dell.
+
+## ARP
+Address Resolution Protocol (ARP) is a protocol used to resolve IP addresses to MAC addresses. For example let's imagine a sender, A, who has MAC `00:DE:AD:BE:EF:00`,  broadcasting a message that essentially asks "Who has IP address `42.42.42.42` please tell A at 00:DE:AD:BE:EF:00". 
+
+If a machine, B, with MAC `12:34:56:78:9a:bc` has the IP address `42.42.42.42` they send a unicast reply back to the sender with the info "`12:34:56:78:9a:bc` has `42.42.42.42`". The sender stores this information in an arp table so whenever it receives packets meant for machine B i.e. a packet with an destination IP address of `42.42.42.42` it sends the packet to MAC it received from B.
```

Comment:
what is unicast vs broadcast in the context of L2?

Reply:
Added more info!